compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1767,7 +1767,24 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                             !needs_infer
                         }
                         Some(_) => true,
-                        None => false,
+                        None => {
+                            if let Some(other) = tcx.impl_trait_ref(other_def)
+                                && let Some(victim) = tcx.impl_trait_ref(victim_def)
+                                && tcx.is_diagnostic_item(sym::AsRef, other.def_id)
+                                && tcx.is_diagnostic_item(sym::AsRef, victim.def_id)
```

Comment:
These would have to be `lang_item` checks instead (to not lie about `AsRef` and `Vec` not being language items after this).